_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+                PublicKeyCredential.parseRegistrationResponseJson(credential);
+            FinishRegistrationOptions options = FinishRegistrationOptions.builder()
+                .request(requestOptions)
+                .response(pkc)
+                .build();
+            RegistrationResult result = relyingParty.finishRegistration(options);
+            Authenticator savedAuth = new Authenticator(result, pkc.getResponse(), user, credname);
+            service.getAuthRepository().save(savedAuth);
+            return new ModelAndView("redirect:/login", HttpStatus.SEE_OTHER);
+        } else {
+            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, 
+                "Cached request expired. Try to register again!");
+        }
+    } catch (RegistrationFailedException e) {
+        throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "Registration failed.", e);
+    } catch (IOException e) {
+        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Failed to save credenital, please try again!", e);
+    }
+}
+
+@PostMapping("/welcome")
+public String finishLogin(@RequestParam String credential,
+                          @RequestParam String username,
+                          Model model,
+                          HttpSession session) {
+    try {
+        PublicKeyCredential<AuthenticatorAssertionResponse, ClientAssertionExtensionOutputs> pkc;
+        pkc = PublicKeyCredential.parseAssertionResponseJson(credential);
+        AssertionRequest request = (AssertionRequest)session.getAttribute(username);
+        AssertionResult result = relyingParty.finishAssertion(FinishAssertionOptions.builder()
+            .request(request)
+            .response(pkc)
+            .build());
+        if (result.isSuccess()) {
+            model.addAttribute("username", username);
+            return "welcome";
+        } else {
+            return "index";
+        }
+    } catch (IOException e) {
+        throw new RuntimeException("Authentication failed", e);
+    } catch (AssertionFailedException e) {
+        throw new RuntimeException("Authentication failed", e);
+    }
+}
+```
+
+> **NOTE:** These requests require server state to work properly. The authentication request includes a challenge nonce generated as part of the `StartRegistrationOptions` and `AssertionRequest` objects. The reply from the client must include this nonce, so the server uses the session to cache the short-lived challenge by caching the full registration (or assertion) request object.
+
+Information is sent to the browser, where it's encoded and passed to the WebAuthn API as a [`PublicKeyCredential` object](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential). The response containing all the data needed for verification is sent to the server and used to create or verify a credential.
+
+#### Registration ceremony
```

Comment:
This heading isn't great. Maybe `Start the registration ceremony`?

Reply:
```suggestion
#### Start the registration ceremony
```